File: BiRetroSys-cpp/Search/src/searchTree.cpp
#include <Search/tree_utils.h>

namespace Search {
    moleculeNode::moleculeNode(const int id, const str &mol, const float value, reactionNode *parent, bool isTerminal): id(id), mol(mol), cost(value){
        this->parent = parent;
        this->children = std::vector<reactionNode*>(0);

        this->hasFound = isTerminal;
        this->isTerminal = isTerminal;
        this->isOpen = !isTerminal;
        this->rn = isTerminal ? 0 : value;
        this->vmt = 0;

        this->gvnode = nullptr;

        if (parent) parent->children.push_back(this);
    }

    void moleculeNode::init(){
        assert(this->isOpen);
        this->isOpen = false;
        this->update();
    }

    void moleculeNode::update(){
        assert(!this->isOpen);
        float newRn = std::numeric_limits<float>::max();
        for (const auto reaction : this->children){
            newRn = std::min(newRn, reaction->rn);
            this->hasFound |= reaction->hasFound;
        }
        float dRn = newRn - this->rn;
        this->rn = newRn;
        if (this->parent) this->parent->update(dRn);
    }

    void moleculeNode::updateVmt(){
        this->vmt = 0;
        if (!this->parent) this->vmt = this->rn;
        else{
            this->vmt += this->parent->costs;
            reactionNode *temp = this->parent;
            while (temp){
                for (const auto mol : temp->children) this->vmt += mol->rn;
                temp = temp->parent->parent;
            }
            for (const auto &reaction : this->children) reaction->updateVmt();
        }
    }

    void moleculeNode::close(){
        this->isOpen = false;
        this->rn = std::numeric_limits<float>::max();
        if (this->parent) this->parent->update(this->rn);
    }

    void moleculeNode::getAncestor(std::unordered_set<str> &ancestor){
        if (this->parent) this->parent->parent->getAncestor(ancestor);
        ancestor.insert(this->mol);
    }

    void moleculeNode::upToDownUpdate(){
        if (!this->parent){
            for (const auto &child : this->children) child->updateVmt();
        }
        else if (!this->parent->parent->parent) this->parent->updateVmt();
        else this->parent->parent->upToDownUpdate();
    }

    void moleculeNode::vizPrepare(Agraph_t *curGraph){
        str nodeName = std::to_string(this->id) + " | " + this->mol;
        this->gvnode = agnode(curGraph, nodeName.data(), 1);
        
        str nodeColor = this->isOpen ? "lightgrey" : "aquamarine";

        if (this->hasFound) nodeColor = "lightblue";
        else if (this->isTerminal) nodeColor = "lightyellow";
        agsafeset(this->gvnode, "color", nodeColor.c_str(), "");
        agsafeset(this->gvnode, "shape", "box", "");
        agsafeset(this->gvnode, "style", "filled", "");
    }

    moleculeNode::~moleculeNode(){}

    reactionNode::reactionNode(const int id, const float value, moleculeNode *parent): id(id), cost(value){
        this->parent = parent;
        this->children = std::vector<moleculeNode*>(0);

        this->hasFound = false;
        this->isOpen = true;
        this->costs = 0;
        this->rn = 0;

        this->gvnode = nullptr;

        if (parent) parent->children.push_back(this);
    }

    float reactionNode::getCosts(){
        return this->parent->parent ? this->cost + this->parent->parent->getCosts() : this->cost;
    }

    void reactionNode::init(){
        assert(this->isOpen);
        this->isOpen = false;

        this->hasFound = true;
        this->rn = 0;
        for (const auto mol : this->children){
            this->rn += mol->rn;
            this->hasFound &= mol->hasFound;
        }
        this->costs = this->getCosts();
    }

    void reactionNode::update(float dRn){
        assert(!this->isOpen);
        this->rn += dRn;
        this->hasFound = true;
        for (const auto mol : this->children) this->hasFound &= mol->hasFound;
        this->parent->update();
    }

    void reactionNode::updateVmt(){
        for (const auto mol : this->children) mol->updateVmt();
    }

    void reactionNode::vizPrepare(Agraph_t *curGraph){
        this->gvnode = agnode(curGraph, std::to_string(this->id).data(), 1);
        
        str nodeColor = this->isOpen ? "lightgrey" : "aquamarine";

        if (this->hasFound) nodeColor = "lightblue";
        agsafeset(this->gvnode, "color", nodeColor.c_str(), "");
        agsafeset(this->gvnode, "shape", "rarrow", "");
        agsafeset(this->gvnode, "style", "filled", "");
    }

    reactionNode::~reactionNode(){}

    // searchTree
    searchTree::searchTree(const str &target, const str &targetName, const std::unordered_set<str> *terminalMol, const int expansionWidth, const int checkWidth, const int singleSteps, const float T): target(target), targetName(targetName), terminalMol(terminalMol), expansionWidth(expansionWidth), checkWidth(checkWidth), singleSteps(singleSteps), T(T){
        this->hasFound = false;
        if (this->terminalMol->find(this->target) != this->terminalMol->end()){
            this->hasFound = true;
            outputLog("Target Molecule already in terminal Molecules.", this->searchLog);
        }

        this->valModel = new valueModel();
        this->inferModel = new Inference::SeqAGraphInfer();
        this->root = this->addMol(this->target, nullptr, this->valueFun({this->target})[0]);
        this->excludeMols = {"", "CC"};

        // load LOG
        str logName = std::filesystem::current_path().parent_path();
        logName += "/Search/" + targetName + "/";
        if (!std::filesystem::exists(logName)) std::filesystem::create_directories(logName);

        this->savePath = logName;
        logName += targetName + ".log";
        this->searchLog.open(logName, std::ios::out | std::ios::trunc);
    }

    std::pair<bool, int> searchTree::multiStepSearch(const int steps, const int earlyStop, const float lowerBound, const bool consistCheck, const float checkLowerBound){
        int step = 0;
        if (!this->hasFound){
            for (; step < steps; step++){
                moleculeNode *nextMol = nullptr;
                float minCost = MAXFLOAT;
                for (auto p : this->molNodes){
                    if (p->isOpen && p->vmt < minCost){
                        minCost = p->vmt;
                        nextMol = p;
                    }
                }

                if (!nextMol){
                    outputLog("No open nodes, search terminate.", this->searchLog);
                    break;
                }
                outputLog("Step " + std::to_string(step+1) + "/" + std::to_string(steps) + ": Trying to expand " + nextMol->mol, this->searchLog);

                auto [expandSmis, expandScores] = this->filterRun(nextMol->mol, this->inferFun({nextMol->mol}), lowerBound, consistCheck, checkLowerBound);

                bool routeFound = this->expandTree(nextMol, expandSmis, expandScores);
                if (routeFound && !this->hasFound){
                    this->hasFound = true;
                    outputLog("Route has found!", this->searchLog);
                }
                if (this->hasFound && step+1 > earlyStop) break;
            }
        }
        return std::make_pair(this->finishSearch(), step+1);
    }

    moleculeNode *searchTree::addMol(const str &mol, reactionNode *parent, float value){
        moleculeNode *newMol = new moleculeNode(this->molNodes.size(), mol, value, parent, this->terminalMol->find(mol) != this->terminalMol->end());
        this->molNodes.push_back(newMol);
        return newMol;
    }

    reactionNode *searchTree::addReaction(const std::vector<str> &reaction, moleculeNode *parent, float cost, std::unordered_set<str> &ancestor){
        for (const auto &mol : reaction){
            if (ancestor.find(mol) != ancestor.end()) return (reactionNode*)nullptr;
        }
        auto values = this->valueFun(reaction);
        reactionNode *newReaction = new reactionNode(this->reacNodes.size(), cost, parent);
        for (int i=0; i < reaction.size(); i++) this->addMol(reaction[i], newReaction, values[i]);
        newReaction->init();
        this->reacNodes.push_back(newReaction);
        return newReaction;
    }

    bool searchTree::expandTree(moleculeNode *startMol, std::vector<std::vector<str>> &expandResults, std::vector<float> &scores){
        assert(!startMol->hasFound);
        if (scores.size() == 0){
            startMol->close();
            return false;
        }
        else{
            std::for_each(scores.begin(), scores.end(), [](float &a){
                a = std::max(std::min(1.0f, a), 1e-3f);
                a = -log(a);
            });
            std::unordered_set<str> ancestor;
            startMol->getAncestor(ancestor);
            for (int i=0; i < scores.size(); i++) this->addReaction(expandResults[i], startMol, scores[i], ancestor);

            if (startMol->children.size() == 0){
                startMol->close();
                return false;
            }
            else{
                startMol->init();
                startMol->upToDownUpdate();
                return this->root->hasFound;
            }
        }
    }

    std::pair<std::vector<std::vector<str>>, std::vector<float>> searchTree::filterRun(const str &expandSmi, const std::vector<std::unordered_map<str, float>> &expandResults, const float lowerBound, const bool consistCheck, const float checkLowerBound){
        std::vector<std::vector<str>> tempRes1;
        std::vector<str> checkInput;
        std::vector<float> tempScore1;
        auto strSplit = [](const str &s, const char split){
            std::vector<str> res;
            int i = 0, startPos = 0;
            for (; i < s.size(); i++){
                if (s[i] == split){
                    if (i > startPos) res.push_back(s.substr(startPos, i - startPos));
                    startPos = i + 1;
                }
            }
            if (i > startPos) res.push_back(s.substr(startPos, i - startPos));
            return res;
        };

        for (auto &[r, s] : expandResults[0]){
            if (s < lowerBound) continue;
            tempRes1.push_back(strSplit(r, '.'));
            tempScore1.push_back(s);
            checkInput.push_back(r);
        }

        // consistency check
        if (consistCheck && checkInput.size()){
            std::vector<std::vector<str>> tempRes2;
            std::vector<float> tempScore2;
            auto checkRes = this->inferFun(checkInput, false);
            int count = 0;
            for (auto &check : checkRes){
                auto findRes = check.find(expandSmi);
                if (findRes != check.end() && findRes->second >= checkLowerBound){
                    tempRes2.push_back(tempRes1[count]);
                    tempScore2.push_back(tempScore1[count] * findRes->second);
                }
                count++;
            }

            // if no results pass consistency check at first step, ignore it
            if (!tempScore2.size() && expandSmi == this->target){
                tempRes2.swap(tempRes1);
                tempScore2.swap(tempScore1);
            }
            if (tempRes2.size() != expandResults[0].size()){
                float accumRes = std::reduce(tempScore2.begin(), tempScore2.end());
                std::for_each(tempScore2.begin(), tempScore2.end(), [accumRes](auto &a){a /= accumRes;});
            }
            return std::make_pair(tempRes2, tempScore2);
        }
        else{
            if (tempScore1.size() != expandResults[0].size()){
                float accumRes = std::reduce(tempScore1.begin(), tempScore1.end());
                std::for_each(tempScore1.begin(), tempScore1.end(), [accumRes](auto &a){a /= accumRes;});
            }
            return std::make_pair(tempRes1, tempScore1);
        }
    }

    std::vector<float> searchTree::valueFun(const std::vector<str> &smis){
        return this->valModel->valueRun(smis);
    }

    std::vector<std::unordered_map<str, float>> searchTree::inferFun(const std::vector<str> &smis, const bool isRetro){
        std::vector<int64_t> lTask = std::vector<int64_t>(smis.size(), isRetro ? 0 : 1);
        int64_t beamSize = isRetro ? this->expansionWidth : this->checkWidth;
        auto [inferRes, inferScore] = this->inferModel->inferRun(smis, lTask, beamSize, smis.size(), 0.0, 1, this->singleSteps, 1, this->T, beamSize); // [batchSize * beamSize]

        for (auto &p : smis) this->excludeMols.insert(p);

        // canonical && filter
        std::vector<std::unordered_map<str, float>> filterRes(smis.size(), std::unordered_map<str, float>());
        for (int i=0; i < inferRes.size(); i++){
            int batchId = i / beamSize;
            auto [canoSmi, isValid] = this->inferModel->molHandler.canonicalizeSmiles(inferRes[i], false);
            if (isValid && this->excludeMols.find(canoSmi) == this->excludeMols.end() && filterRes[batchId].find(canoSmi) == filterRes[batchId].end()) filterRes[batchId][canoSmi] = inferScore[i];
        }
        for (auto &m : filterRes){
            std::for_each(m.begin(), m.end(), [](auto &p){p.second = exp(p.second);});
            float scoreSum = std::reduce(m.begin(), m.end(), 0.0f, [](float curSum, const auto &p){return curSum + p.second;});
            std::for_each(m.begin(), m.end(), [&scoreSum](auto &p){p.second /= scoreSum;});
        }
        for (auto &p : smis) this->excludeMols.erase(p);
        return filterRes;
    }

    void searchTree::visualization(const str &name){
        GVC_t *gvc = gvContext();
        Agraph_t *G = agopen("g", Agdirected, 0);

        std::queue<std::pair<std::variant<moleculeNode*, reactionNode*>, std::variant<moleculeNode*, reactionNode*>>> qNode;
        reactionNode *tempNode = nullptr;
        this->root->vizPrepare(G);
        qNode.push(std::make_pair(this->root, tempNode));

        while (!qNode.empty()){
            auto [__node, __parent] = qNode.front(); qNode.pop();
            // reactionNode -> moleculeNode
            if (std::get_if<moleculeNode*>(&__node)){
                auto node = std::get<moleculeNode*>(__node);
                auto parent = std::get<reactionNode*>(__parent);
                if (parent) auto bond = agedge(G, parent->gvnode, node->gvnode, "", 1);
                if (node->children.size() > 0){
                    for (auto &p : node->children){
                        p->vizPrepare(G);
                        qNode.push(std::make_pair(p, node));
                    }
                }
            }
            else if (std::get_if<reactionNode*>(&__node)){
                // moleculeNode -> reactionNode
                auto node = std::get<reactionNode*>(__node);
                auto parent = std::get<moleculeNode*>(__parent);
                str bondLabel = std::to_string(exp(-node->cost));
                bondLabel = bondLabel.substr(0, bondLabel.find('.') + 3);
                auto bond = agedge(G, parent->gvnode, node->gvnode, bondLabel.data(), 1);
                if (node->children.size() > 0){
                    for (auto &p : node->children){
                        p->vizPrepare(G);
                        qNode.push(std::make_pair(p, node));
                    }
                }
            }
        }
        agsafeset(G, "dpi", "600", "");
        agsafeset(G, "rankdir", "LR", "");
        gvLayout(gvc, G, "dot");
        gvRenderFilename(gvc, G, "pdf", (name + ".pdf").c_str());
        gvFreeLayout(gvc, G);
        agclose(G);
        gvFreeContext(gvc);
    }

    void searchTree::visualizationBest(const str &name){
        GVC_t *gvc = gvContext();
        Agraph_t *G = agopen("g", Agdirected, 0);

        std::queue<moleculeNode*> qNode;
        this->root->vizPrepare(G);
        qNode.push(this->root);

        while (!qNode.empty()){
            auto node = qNode.front(); qNode.pop();
            if (node->hasFound){
                reactionNode *bestReaction = nullptr;
                if (node->children.size() > 0){
                    for (auto p : node->children){
                        if (p->hasFound && (!bestReaction || p->cost < bestReaction->cost)) bestReaction = p;
                    }
                    str bondLabel = std::to_string(exp(-bestReaction->cost));
                    bondLabel = bondLabel.substr(0, bondLabel.find('.') + 3);
                    bestReaction->vizPrepare(G);
                    auto bond = agedge(G, node->gvnode, bestReaction->gvnode, bondLabel.data(), 1);

                    for (auto p : bestReaction->children){
                        p->vizPrepare(G);
                        auto chbond = agedge(G, bestReaction->gvnode, p->gvnode, "", 1);
                        qNode.push(p);
                    }
                }
            }
        }
        agsafeset(G, "dpi", "600", "");
        agsafeset(G, "rankdir", "LR", "");
        gvLayout(gvc, G, "dot");
        gvRenderFilename(gvc, G, "pdf", (name + ".pdf").c_str());
        gvFreeLayout(gvc, G);
        agclose(G);
        gvFreeContext(gvc);
    }

    bool searchTree::finishSearch(){
        this->visualization(this->savePath + this->targetName + "-Tree");
        if (this->hasFound) this->visualizationBest(this->savePath + this->targetName + "-BestRoute");
        return this->hasFound;
    }

    searchTree::~searchTree(){
        this->searchLog.close();
        for (auto p : this->molNodes) delete p;
        for (auto p : this->reacNodes) delete p;
        std::vector<moleculeNode*>().swap(this->molNodes);
        std::vector<reactionNode*>().swap(this->reacNodes);

        delete this->valModel;
        delete this->inferModel;
    }

    void loadTerminalMols(std::unordered_set<str> &finalSet, str &&molPath){
        if (molPath == "" || !std::filesystem::exists(molPath)){
            molPath = std::filesystem::current_path().parent_path();
            molPath += "/Models/origin_dict.csv";
        }
        auto loadBegin = std::chrono::high_resolution_clock::now();
        
        #if (defined _OPENMP) && (_RDPAL)
        #pragma omp parallel
        {
            std::ifstream terminalMols;
            terminalMols.open(molPath, std::ios::in);
            std::unordered_set<str> localSet;
            str line;
            std::getline(terminalMols, line);

            int tid = omp_get_thread_num();
            int tnum = omp_get_num_threads();
            
            // skip
            for (int i=0; i < tid; i++) std::getline(terminalMols, line);
            while (std::getline(terminalMols, line)){
                int start = 1, end=line.size() - 1;
                for (; start < end && line[end] != ','; end--){}
                if (start < end) localSet.insert(line.substr(start+1, end-start-1));
                // skip
                for (int i=0; i < tnum && std::getline(terminalMols, line); i++){}
            }

            // accumulate
            #pragma omp critical
            {
                finalSet.insert(localSet.begin(), localSet.end());
            }
            terminalMols.close();
        }
        #else
        std::ifstream terminalMols;
        terminalMols.open(molPath, std::ios::in);
        str line;
        std::getline(terminalMols, line);

        while (std::getline(terminalMols, line)){
            int start = 0, end=line.size() - 1;
            for (; start < end && line[start] != ','; start++){}
            for (; start < end && line[end] != ','; end--){}
            if (start < end) finalSet.insert(line.substr(start+1, end-start-1));
        }
        terminalMols.close();
        #endif

        auto loadEnd = std::chrono::high_resolution_clock::now();
        auto loadCost = std::chrono::duration_cast<std::chrono::milliseconds>(loadEnd - loadBegin).count() * 1e-3;
        printf("load %lu molecules from %s, cost %.2f second\n", finalSet.size(), molPath.data(), loadCost);
    }
}

File: BiRetroSys-py/MultiStepSearch/searchTree.py
import os
import logging
import math
import numpy as np

from tqdm.std import trange
from MultiStepSearch.moleculeNode import molNode
from MultiStepSearch.reactionNode import reactionNode
from queue import Queue
from graphviz import Digraph

curDir = os.path.dirname(os.path.realpath(__file__))
saveDir = os.path.join(curDir, "SearchResults")

class molTree:
    def __init__(self, targrtMol: str, terminalMol: list[str], valueFun, inferFun, checkFun)-> None:
        self.targetMol = targrtMol
        self.terminalMol = terminalMol
        self.valueFun = valueFun
        self.inferFun = inferFun
        self.checkFun = checkFun

        self.molNodes = []
        self.reactionNodes = []
        self.hasFound = targrtMol in terminalMol

        if self.hasFound:
            logging.info("Target Molecule already in terminal Molecules.")
        
        self.root = self.__addMol(targrtMol, None, list(self.valueFun([targrtMol]))[0])

    def __addMol(self, mol: str, parent: reactionNode, value: float):
        hasFound = mol in self.terminalMol

        mNode = molNode(
            mol=mol,
            value=value,
            parent=parent,
            molId=len(self.molNodes),
            isTerminal=hasFound
        )
        self.molNodes.append(mNode)
        return mNode
    
    def __addReaction(self, reaction: list[str], parent: molNode, cost: float, ancestor: set):
        for mol in reaction:
            if mol in ancestor: return None
        
        values = list(self.valueFun(reaction))
        rNode = reactionNode(cost, parent, len(self.reactionNodes))
        for mol, value in zip(reaction, values):
            self.__addMol(mol, rNode, value)
        
        rNode.initValue()
        self.reactionNodes.append(rNode)
        return rNode

    def __expand(self, startMol: molNode, smiResult: list[list[str]], scores: list[float]):
        assert not startMol.hasFound and len(startMol.children) == 0

        if len(scores) == 0:
            startMol.closeNode()
            return False
        
        scores = 0.0 - np.log(np.clip(np.array(scores), 1e-3, 1.0))
        scores = list(scores)
        for i in range(len(scores)):
            self.__addReaction(smiResult[i], startMol, scores[i], startMol.getAncestor())
        
        if len(startMol.children) == 0:
            startMol.closeNode()
            return False
        
        startMol.initValue(propagate=True)
        startMol.firstReactionUpdateVmt()
        return self.root.hasFound
    
    #graphviz visualization with BFS
    def __visualization(self, name: str):
        G = Digraph("searchTree", filename=os.path.join(saveDir, name))
        G.attr(rankdir="LR")
        G.attr("node", shape="box")
        G.format = "pdf"

        qNode = Queue()
        qNode.put((self.root, None)) #node & parent
        while not qNode.empty():
            node, parent = qNode.get()
            color = "lightgrey" if node.isOpen else "aquamarine"
            shape = "box" if isinstance(node, molNode) else "rarrow"

            if isinstance(node, molNode) and node.isTerminal:
                color = "lightyellow"
            elif node.hasFound:
                color = "lightblue"
            
            G.node(node.getStrFeature(), shape=shape, color=color, style="filled")

            label = "{:.4f}".format(math.exp(-node.cost)) if isinstance(node, reactionNode) else ""
            if parent is not None:
                G.edge(parent.getStrFeature(), node.getStrFeature(), label=label)
            if len(node.children) > 0:
                for child in node.children:
                    qNode.put((child, node))
        G.render()
    
    def __visualizationBest(self, name: str):
        G = Digraph("searchTree", filename=os.path.join(saveDir, name))
        G.attr(rankdir="LR")
        G.attr("node", shape="box")
        G.format = "pdf"

        color, terminalColor = "lightblue", "lightyellow"
        molShape, reactionShape = "box", "rarrow"

        qNode = Queue()
        qNode.put(self.root) #node & parent
        G.node(self.root.getStrFeature(), shape=molShape, color=terminalColor if self.root.isTerminal else color, style="filled")
        
        while True:
            mol = qNode.get()
            if mol.hasFound:
                bestReaction = None
                if len(mol.children) > 0:
                    for reaction in mol.children:
                        if (reaction.hasFound) and (bestReaction is None or reaction.cost < bestReaction.cost):
                            bestReaction = reaction

                    label = "{:.4f}".format(math.exp(-bestReaction.cost))
                    G.node(bestReaction.getStrFeature(), shape=reactionShape, color=color, style="filled")
                    G.edge(mol.getStrFeature(), bestReaction.getStrFeature(), label=label)
                
                    for nextmol in bestReaction.children:
                        qNode.put(nextmol)
                        G.node(nextmol.getStrFeature(), shape=molShape, color=terminalColor if nextmol.isTerminal else color, style="filled")
                        G.edge(bestReaction.getStrFeature(), nextmol.getStrFeature(), label="")

            if qNode.empty(): break
        G.render()
    
    def __finish(self, name: str="Tree", bestname: str="BestRoute"):
        if not os.path.exists(saveDir): os.makedirs(saveDir)

        self.__visualization(name)
        if self.hasFound:
            self.__visualizationBest(bestname)

        return self.hasFound

    def search(self, steps: int, name: str="Tree", bestName: str="bestRoute", earlyStop: int=0, lowerbound: float=0.1, consistCheck: bool=True, checkLowerBound: float=0.01):
        if not self.hasFound:
            tqdmSteps = trange(steps)
            for step in tqdmSteps:
                scores = []
                for i in range(len(self.molNodes)):
                    scores.append(self.molNodes[i].vmt) if self.molNodes[i].isOpen else scores.append(float("inf"))
                
                if min(scores) == float("inf"):
                    logging.info("No open nodes!")
                    break

                nextNodeIdx = scores.index(min(scores))
                nextNode = self.molNodes[nextNodeIdx]

                logging.info(f"Step{step}: Trying to expand {nextNode.mol}")
                assert nextNode.isOpen

                inferRes, inferScore = self.inferFun([nextNode.mol], nextNode.mol)

                filterMap = ("CC", "")

                #hard-coded filtering
                #1. filter invalid results and rescale
                filterRes, filterScore = [], []
                for smiRes, smiScore in zip(inferRes[0], inferScore[0]):
                    if smiRes in filterMap: continue
                    filterRes.append(smiRes)
                    filterScore.append(smiScore)
                scoreSum = sum(filterScore)
                filterScore = [_ / scoreSum for _ in filterScore]

                #2. split reactants & lowerbound
                splitRes, splitScore, checkInput = [], [], []
                for smiRes, smiScore in zip(filterRes, filterScore):
                    reactantSplit = smiRes.split(".")
                    if smiScore < lowerbound: continue

                    splitRes.append(reactantSplit)
                    splitScore.append(smiScore)
                    checkInput.append(smiRes)
                
                #3. consistency check
                if consistCheck and len(checkInput) > 0:
                    consistRes, consistScore = [], []
                    checkRes, checkScore = self.checkFun(checkInput, "")
                    for i, (res, score) in enumerate(zip(checkRes, checkScore)):
                        if nextNode.mol in res and score[res.index(nextNode.mol)] >= checkLowerBound:
                            idx = res.index(nextNode.mol)
                            consistRes.append(splitRes[i])
                            consistScore.append(splitScore[i] * score[idx])
                
                    #rescale
                    if len(consistScore) == 0 and nextNode == self.root:
                        consistRes = splitRes
                        consistScore = splitScore
                        logging.info(f"{nextNode.mol} no results pass consistency check, now trying more flexible settings")

                    if len(consistScore) != len(inferScore[0]):
                        scoreSum = sum(consistScore)
                        consistScore = [score / scoreSum for score in consistScore]
                    
                    hasFound = self.__expand(nextNode, consistRes, consistScore)
                
                else:
                    #rescale
                    if len(splitScore) != len(inferScore[0]):
                        scoreSum = sum(splitScore)
                        splitScore = [score / scoreSum for score in splitScore]
                    
                    hasFound = self.__expand(nextNode, splitRes, splitScore)

                if hasFound and not self.hasFound:
                    self.hasFound = hasFound
                    logging.info("Route has found!")
                
                if self.hasFound and step + 1 > earlyStop:
                    break
                
        return self.__finish(name, bestName), step + 1